File: backend/app/core/period.py
"""
Résolution des périodes de filtrage — SOURCE UNIQUE DE VÉRITÉ.

Tout module qui filtre des données par date (Tableau de bord, Finance,
Analytics, exports…) DOIT passer par `resolve_period()`.

C'est cette fonction qui garantit qu'une même sélection de dates produit
exactement la même fenêtre temporelle dans tous les modules, et donc
exactement les mêmes chiffres.

Conventions retenues :
- Toutes les dates renvoyées sont *timezone-aware* en UTC.
- La borne de début est **inclusive**, la borne de fin est **exclusive**
  (`start <= created_at < end`). L'appelant n'a donc jamais à gérer
  lui-même le « +1 jour » pour inclure la journée de fin.
- Une date explicite (`start_date` / `end_date`) l'emporte toujours sur
  le libellé de période : c'est ce qui permet au Tableau de bord et à
  Finance de rester synchronisés même si leurs sélecteurs diffèrent.
- Les périodes glissantes sont ancrées sur des frontières de journée
  (minuit UTC) et non sur l'instant présent : deux requêtes envoyées à
  quelques secondes d'intervalle renvoient ainsi des résultats identiques.
"""
from __future__ import annotations

from datetime import datetime, timedelta, timezone

# Libellés de période acceptés par l'ensemble de l'API.
PERIOD_VALUES = ("24h", "7j", "30j", "90j", "all", "custom")

# Nombre de jours associé à chaque période glissante.
_PERIOD_DAYS = {"7j": 7, "30j": 30, "90j": 90}


def _parse_date(value: str | None) -> datetime | None:
    """
    Convertit une date fournie par le client en datetime UTC.

    Accepte `YYYY-MM-DD` ainsi que les formats ISO complets renvoyés par
    `Date.toISOString()` côté navigateur. Renvoie None si la valeur est
    vide ou illisible (le filtre est alors simplement ignoré).
    """
    if not value:
        return None

    raw = value.strip()
    if not raw:
        return None

    try:
        # `fromisoformat` gère "2026-07-25" comme "2026-07-25T10:30:00+00:00".
        parsed = datetime.fromisoformat(raw.replace("Z", "+00:00"))
    except ValueError:
        return None

    if parsed.tzinfo is None:
        return parsed.replace(tzinfo=timezone.utc)
    return parsed.astimezone(timezone.utc)


def _start_of_day(moment: datetime) -> datetime:
    return moment.replace(hour=0, minute=0, second=0, microsecond=0)


def resolve_period(
    period: str | None = None,
    start_date: str | None = None,
    end_date: str | None = None,
    now: datetime | None = None,
) -> tuple[datetime | None, datetime | None]:
    """
    Traduit les paramètres de filtrage en fenêtre `[start, end)` UTC.

    Renvoie `(None, None)` pour « toutes les données », ce qui permet à
    l'appelant de ne poser aucune contrainte de date sur sa requête.

    Règles de priorité :
    1. `start_date` et/ou `end_date` fournies  → fenêtre explicite.
    2. `period` = "all"                        → aucune contrainte.
    3. `period` = "24h" / "7j" / "30j" / "90j" → fenêtre glissante.
    4. Rien de reconnaissable                  → aucune contrainte.
    """
    current = now.astimezone(timezone.utc) if now else datetime.now(timezone.utc)

    explicit_start = _parse_date(start_date)
    explicit_end = _parse_date(end_date)

    # ── 1. Dates explicites : elles priment sur le libellé de période ──
    if explicit_start or explicit_end:
        start = _start_of_day(explicit_start) if explicit_start else None
        end = None
        if explicit_end:
            # Borne haute exclusive : on inclut la totalité de la journée de fin.
            end = _start_of_day(explicit_end) + timedelta(days=1)

        # Une plage inversée serait vide et donnerait des KPI à zéro sans
        # explication : on la remet à l'endroit plutôt que de piéger l'utilisateur.
        if start and end and start >= end:
            start, end = _start_of_day(end - timedelta(days=1)), start + timedelta(days=1)

        return start, end

    # ── 2. Aucune contrainte ──
    if period == "all":
        return None, None

    # ── 3. Périodes glissantes ──
    if period == "24h":
        return current - timedelta(hours=24), current

    days = _PERIOD_DAYS.get(period or "")
    if days is not None:
        # Ancrage sur minuit : la fenêtre couvre les N derniers jours pleins
        # plus la journée en cours, exactement comme le sélecteur du frontend.
        start = _start_of_day(current) - timedelta(days=days)
        end = _start_of_day(current) + timedelta(days=1)
        return start, end

    # ── 4. Valeur inconnue : on ne filtre pas plutôt que de mentir ──
    return None, None


def previous_period(
    start: datetime | None,
    end: datetime | None,
) -> tuple[datetime | None, datetime | None]:
    """
    Fenêtre immédiatement antérieure et de même durée, utilisée pour les
    variations en pourcentage (« +12,4 % sur la période »).

    Renvoie `(None, None)` si la période courante n'est pas bornée : on ne
    peut pas comparer « tout l'historique » à quoi que ce soit.
    """
    if start is None or end is None:
        return None, None

    span = end - start
    return start - span, start

File: frontend/src/lib/period.ts
/**
 * Périodes de filtrage — SOURCE UNIQUE DE VÉRITÉ côté client.
 *
 * Tout écran qui propose un filtre par date (Tableau de bord, Finance,
 * Analytics) doit produire ses paramètres avec `buildPeriodParams()`.
 * Deux écrans qui affichent la même période envoient ainsi exactement les
 * mêmes `start_date` / `end_date` au backend, lequel les interprète avec
 * la fonction jumelle `app/core/period.py::resolve_period()`.
 */

export type PeriodKey = '24h' | '7j' | '30j' | '90j' | 'all' | 'custom';

export interface PeriodParams {
  period: PeriodKey;
  start_date?: string;
  end_date?: string;
}

/** Nombre de jours couvert par chaque période glissante. */
const PERIOD_DAYS: Partial<Record<PeriodKey, number>> = {
  '7j': 7,
  '30j': 30,
  '90j': 90,
};

/**
 * Formate une date en `YYYY-MM-DD` à partir de ses composantes **locales**.
 *
 * `toISOString()` convertit d'abord en UTC : passé 00 h dans un fuseau en
 * avance, il renvoie la veille et décale toute la période d'un jour.
 */
export function toDateInput(date: Date): string {
  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, '0');
  const day = String(date.getDate()).padStart(2, '0');
  return `${year}-${month}-${day}`;
}

/** Date d'aujourd'hui au format attendu par `<input type="date">`. */
export function today(): string {
  return toDateInput(new Date());
}

/**
 * Traduit une sélection d'interface en paramètres d'API.
 *
 * Pour une période personnalisée incomplète (une seule des deux dates
 * saisies), la borne renseignée est tout de même transmise : l'utilisateur
 * voit immédiatement l'effet de sa saisie.
 */
export function buildPeriodParams(
  period: PeriodKey,
  customStart?: string,
  customEnd?: string,
): PeriodParams {
  if (period === 'custom') {
    return {
      period: 'custom',
      start_date: customStart || undefined,
      end_date: customEnd || undefined,
    };
  }

  if (period === 'all' || period === '24h') {
    // "all" ne porte aucune borne ; "24h" est une fenêtre horaire que seul
    // le backend peut ancrer précisément.
    return { period };
  }

  const days = PERIOD_DAYS[period];
  if (!days) return { period };

  const start = new Date();
  start.setDate(start.getDate() - days);

  return {
    period,
    start_date: toDateInput(start),
    end_date: today(),
  };
}

/**
 * Teste si une date ISO tombe dans la période sélectionnée.
 *
 * Utilisé pour les listes filtrées côté client (commandes récentes) afin
 * qu'elles couvrent exactement la même fenêtre que les KPI : borne de début
 * inclusive, journée de fin incluse en entier.
 */
export function isWithinPeriod(
  isoDate: string | null | undefined,
  params: PeriodParams,
): boolean {
  if (!isoDate) return false;
  if (!params.start_date && !params.end_date) return true;

  const time = new Date(isoDate).getTime();
  if (Number.isNaN(time)) return false;

  if (params.start_date) {
    const from = new Date(`${params.start_date}T00:00:00`).getTime();
    if (time < from) return false;
  }
  if (params.end_date) {
    const to = new Date(`${params.end_date}T23:59:59.999`).getTime();
    if (time > to) return false;
  }
  return true;
}

/** Libellés affichés dans les sélecteurs de période. */
export function periodLabel(period: PeriodKey, language: string): string {
  const fr: Record<PeriodKey, string> = {
    '24h': '24 dernières heures',
    '7j': '7 derniers jours',
    '30j': '30 derniers jours',
    '90j': '90 derniers jours',
    all: 'Tout l’historique',
    custom: 'Période personnalisée',
  };
  const en: Record<PeriodKey, string> = {
    '24h': 'Last 24 hours',
    '7j': 'Last 7 days',
    '30j': 'Last 30 days',
    '90j': 'Last 90 days',
    all: 'All time',
    custom: 'Custom range',
  };
  return language === 'fr' ? fr[period] : en[period];
}
